### PR TITLE
ARROW-8581: [C#] Accept and return DateTime from DateXXArray

### PR DESCRIPTION
**Pull request by mr-smidge**
This PR introduces additions to the public API for `Date32Array` and `Date64Array` by accepting and returning `System.DateTime` in additio to `System.DateTimeOffset`.

The rationale for making this change is explained in detail on the [JIRA ticket](https://issues.apache.org/jira/browse/ARROW-8581), but briefly: making this change avoids a class of bugs that manifest based on the user's timezone, and where it is very easy to unknowingly fall into a trap.

Note that the array builders no longer derive from `PrimitiveArrayBuilder<TFrom, TTo, ...>`: that base class is designed for conversion from exactly one outward-facing type to the underlying type, but these builders now have two outward-facing types.  As a replacement, I have introduced abstract types `DateArrayBuilder` (specific to dates) and `DelegatingArrayBuilder` (more general-purpose).   See inline comments for further details.

### CODE REVIEW

**Comment by eerhardt** [comment]
Can we just introduce new `Append(DateTime value)` and `GetDateAsDateTime(int index)` APIs instead of making a breaking change?

In general, DateTimeOffset is the preferred type to use when talking about dates and times. The reasoning is because exactly what you point out in the JIRA issue - DateTime.Kind is very confusing and can cause a lot of problems. So we generally recommend users to use DateTimeOffset unless they have a strong reason to use DateTime.

So I'd suggest to keep the existing code (the builder is based on DateTimeOffset), but to also add new `Append(DateTime value)` overloads to solve the issue you logged.

**Comment by mr-smidge** [comment]
> In general, DateTimeOffset is the preferred type to use when talking about dates and times. The reasoning is because exactly what you point out in the JIRA issue - DateTime.Kind is very confusing and can cause a lot of problems.

I suspect this is a matter of codebase/team preference.  By contract, my organisation previously recommended `DateTime` for dates.  I say "previously" because now we would use NodaTime's `LocalDate` instead :-).

Nonetheless, in the absence of a proper type just for dates, I can see arguments for either, and I'll modify the API accordingly to accept both in a non-breaking manner.

File: csharp/src/Apache.Arrow/Arrays/Date32Array.cs
Comment (mr-smidge) [inline csharp/src/Apache.Arrow/Arrays/Date32Array.cs:86]: In the absence of being able to overload return types, this class (and `Date64Array`) follows the naming convention of [System.Convert](https://docs.microsoft.com/en-us/dotnet/api/system.convert?view=netcore-3.1), and mentions the return type in the method name.

As such, the original `GetDate()` method is marked as _obsolete_.

File: csharp/src/Apache.Arrow/Arrays/DateArrayBuilder.cs
Comment (mr-smidge) [inline csharp/src/Apache.Arrow/Arrays/DateArrayBuilder.cs:21]: This new class is a bit like splitting `PrimitiveArrayBuilder<TFrom, TTo, ...>` into two: one class whose interface is to do with accepting values of differing types for conversion, and another (`DelegatingArrayBuilder`) which simply delegates to an 'inner' array builder for methods like `Resize()` that don't care about the type of value.

For clarity:  `PrimitiveArrayBuilder<TFrom, TTo, ...>` would no longer work as a base class for `DateXXArray.Builder` because the builder now accepts two different kinds of type for writing to the array.

File: csharp/src/Apache.Arrow/Arrays/DelegatingArrayBuilder.cs
Comment (mr-smidge) [inline csharp/src/Apache.Arrow/Arrays/DelegatingArrayBuilder.cs:26]: Perhaps  `PrimitiveArrayBuilder<TFrom, TTo, ...>`  could now use this as a base class?

The issue is that the only remaining user of  `PrimitiveArrayBuilder<TFrom, TTo, ...>`  is now `TimestampArray`, and that class has a number of problems that mean it may need to be refactored to derive from a different base class (in short: it needs to understand IANA/Olson timezone identifiers, and NodaTime is the obvious library to do that, but if we include NodaTime then the class might as well accept `NodaTime.Instant`, etc).  This made me think that  `PrimitiveArrayBuilder<TFrom, TTo, ...>` should probably be marked obsolete.

Opinions welcomed!

File: csharp/test/Apache.Arrow.Tests/Date32ArrayTests.cs
Comment (eerhardt) [inline csharp/test/Apache.Arrow.Tests/Date32ArrayTests.cs:23]: The `GetDate` method is now obsolete. So this should probably be updated to the new method name.

File: csharp/test/Apache.Arrow.Tests/Date32ArrayTests.cs
Comment (eerhardt) [inline csharp/test/Apache.Arrow.Tests/Date32ArrayTests.cs:34]: Can we also get tests for `GetDateTimeOffset`?

File: csharp/src/Apache.Arrow/Arrays/DelegatingArrayBuilder.cs
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Arrays/DelegatingArrayBuilder.cs:32]: Do we have to have this extra layer of abstraction? If we don't think TimestampArray could use this base class, then maybe let's wait with this extra base class. We can always refactor it out of `DateArrayBuilder` later, when it provides value.

File: csharp/src/Apache.Arrow/Arrays/Date64Array.cs
Comment (eerhardt) [inline csharp/src/Apache.Arrow/Arrays/Date64Array.cs:61]: I believe this should stay `value.ToUnixTimeMilliseconds()`. Same for Date32 above.

File: csharp/src/Apache.Arrow/Arrays/Date64Array.cs
Comment (eerhardt) [inline csharp/src/Apache.Arrow/Arrays/Date64Array.cs:104]: I believe this should stay `return DateTimeOffset.FromUnixTimeMilliseconds(value.Value);`. Same for Date32 above.

File: csharp/src/Apache.Arrow/Arrays/Date64Array.cs
Comment (mr-smidge) [reply csharp/src/Apache.Arrow/Arrays/Date64Array.cs]: I've restored this to use `.ToUnixTimeMilliseconds()` (and with the conversion in the other direction).

However, I'm not sure whether it is such a good idea to make the same change in `Date32Array`.  My personal opinion is that `(dateTime - _epoch).TotalDays` is more readable as a day count than having to make a trip into milliseconds and then convert again.  Involving milliseconds may also distract the reader into momentarily thinking about leap seconds, when they are not relevant for a day count.

File: csharp/src/Apache.Arrow/Arrays/Date64Array.cs
Comment (mr-smidge) [reply csharp/src/Apache.Arrow/Arrays/Date64Array.cs]: Restored to use `.FromUnixTimeMilliseconds()`.

File: csharp/test/Apache.Arrow.Tests/Date32ArrayTests.cs
Comment (mr-smidge) [reply csharp/test/Apache.Arrow.Tests/Date32ArrayTests.cs:34]: Added.

File: csharp/src/Apache.Arrow/Arrays/DelegatingArrayBuilder.cs
Comment (mr-smidge) [reply csharp/src/Apache.Arrow/Arrays/DelegatingArrayBuilder.cs:32]: If I were to refactor `TimestampArray` today, I would use `DelegatingArrayBuilder` as the builder's base class, adding methods to accept `DateTimeOffset` (and probably `NodaTime.Instant` if we use NodaTime for the TZDB).

I don't mind which direction we take for this PR, but have a minor preference to keep `DelegatingArrayBuilder` (because we know that we will need to refactor `TimestampArray` in order for the C# codebase to become Production-ready).

I've raised https://issues.apache.org/jira/browse/ARROW-9515 for the enhancements to `TimestampArray` in future.

**Comment by mr-smidge** [comment]
@eerhardt , is there a way to re-trigger build checks?  The `C# / AMD64 MacOS 10.15 C# 2.2.103 (pull_request)` check failed with an unhelpful "This check failed" error message - I assume this is transient?

Thanks!

**Comment by eerhardt** [comment]
> is there a way to re-trigger build checks?

In other repos I've used the "Ru-run jobs" button on the "Checks" tab:

image

File: csharp/src/Apache.Arrow/Arrays/Date64Array.cs
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Arrays/Date64Array.cs]: > However, I'm not sure whether it is such a good idea to make the same change in Date32Array. My personal opinion is that (dateTime - _epoch).TotalDays is more readable as a day count than having to make a trip into milliseconds and then convert again.

This isn't about read-ability, but instead about correctness. The logic needs to take into account the Offset portion of the `DateTimeOffset`.

The Unix Epoch is defined as `00:00:00 UTC Thursday, 1 January 1970`. At that specific time, half of the world was on the day `1 January 1970`, and the other half of the world was still on `31 December 1969`. I can have the same **moment in time** represented with 2 different DateTimeOffsets - one with a negative Offset, and one with a positive Offset. If used the proposed `Date32Array.Builder`, I would get 2 different days. This is not correct.

For example, the following test should pass, but it doesn't:

```C#
            [Fact]
            public void RespectsOffset()
            {
                // Arrange
                var westernHemisphereTime = new DateTimeOffset(2020, 7, 20, 20, 0, 0, TimeSpan.FromHours(-5));
                var easternHemisphereTime = new DateTimeOffset(2020, 7, 21, 10, 0, 0, TimeSpan.FromHours(9));
                var array = new Date32Array.Builder()
                    .Append(westernHemisphereTime)
                    .Append(easternHemisphereTime)
                    .Build();

                // Act
                var actualFirst = array.GetDate(0);
                var actualSecond = array.GetDate(1);

                // Assert
                Assert.NotNull(actualFirst);
                Assert.NotNull(actualSecond);
                Assert.Equal(actualFirst, actualSecond);
                Assert.Equal(new DateTimeOffset(2020, 7, 21, 0, 0, 0, TimeSpan.Zero), actualFirst);
            }
```

File: csharp/src/Apache.Arrow/Arrays/DelegatingArrayBuilder.cs
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Arrays/DelegatingArrayBuilder.cs:32]: > and probably NodaTime.Instant if we use NodaTime for the TZDB

Just a FYI, see this [previous discussion](https://github.com/apache/arrow/pull/5965#discussion_r356233161) where we discussed NodaTime. In short, I don't believe the Arrow library should have a dependency on NodaTime. But instead, the Arrow API should allow for a consumer to use NodaTime and Arrow together.

**Comment by kou** [comment]
I think that a member of the "arrow-committers" team can use "Re-run jobs". (I re-ran jobs.)

Others can use `git commit --amend && git push --force ...` or ask a committer to use "Re-run jobs".

**Comment by eerhardt** [comment]
> I think that a member of the "arrow-committers" team can use "Re-run jobs". (I re-ran jobs.)

Ah, I didn't know committers were special that way. Thanks for re-running, and thanks for the info.

File: csharp/src/Apache.Arrow/Arrays/Date64Array.cs
Comment (mr-smidge) [reply csharp/src/Apache.Arrow/Arrays/Date64Array.cs]: > The Unix Epoch is defined as 00:00:00 UTC Thursday, 1 January 1970.

Apologies, I've been too colloquial with use of the word epoch when talking about dates.  You are right that the UNIX epoch is a point in time, not a date.

However, `Date32` and `Date64` are measures of _dates_.  As the standard C# library does not have a dedicated type for dates, we have to infer a date from either `DateTime` or `DateTimeOffset`.  Dates and instants in time aren't implicitly convertible in the natural sense because of timezones: there has to be some logic in the conversion.

>  The logic needs to take into account the Offset portion of the DateTimeOffset.
> I can have the same moment in time represented with 2 different DateTimeOffsets

Absolutely right - the same point in time can lie within two different calendar dates depending on one's timezone.  At the time I'm writing this it's just gone midnight in the UK, so I think it's 21st July right now.  But if you're on CDT, it's 20th July for you at the very same instant.

For clarity: I think the most natural way to infer a meaningful date from `DateTime`/`DateTimeOffset` is to call the `.Date` property, i.e. I would expect the time and offset to be _ignored_ when considering the date.

> I can have the same moment in time represented with 2 different DateTimeOffsets - one with a negative Offset, and one with a positive Offset. If used the proposed Date32Array.Builder, I would get 2 different days. This is not correct.

I disagree with the statement that this is not correct, but I want to understand your viewpoint better...

From looking at your test case, do you mean to suggest that you believe a conversion to UTC first and then taking the date from the UTC date/time is the most natural way to infer a date from a `DateTimeOffset`?  This is what happens by doing `(int)(dto.ToUnixTimeMilliseconds() / MillisecondsPerDay)`, but it's an interpretation of dates from times that I've never seen in any other project!

If you take the "simplest approach" and call `.Date` (or its equivalent in NodaTime), time and offset are always ignored:

```csharp
// Using System types
var westernHemisphereTime = new DateTimeOffset(2020, 7, 20, 20, 0, 0, TimeSpan.FromHours(-5));
var easternHemisphereTime = new DateTimeOffset(2020, 7, 21, 10, 0, 0, TimeSpan.FromHours(9));
Assert.NotEqual(westernHemisphereTime.Date == easternHemisphereTime.Date);
Assert.Equal(westernHemisphereTime.UtcTicks == easternHemisphereTime.UtcTicks);

// Using NodaTime types
var westernHemisphereZtd = new LocalDateTime(2020, 7, 20, 20, 0, 0).InZoneStrictly(DateTimeZoneProviders.Tzdb["America/Chicago"]);
var easternHemisphereZtd = new LocalDateTime(2020, 7, 21, 10, 0, 0).InZoneStrictly(DateTimeZoneProviders.Tzdb["Asia/Tokyo"]);
Assert.NotEqual(westernHemisphereZtd.Date == easternHemisphereZtd.Date);
Assert.Equal(westernHemisphereZtd.ToInstant() == easternHemisphereZtd.ToInstant());
```

Separately, I've realised that `Date64Array` had an existing bug that I would not be fixing if I were to retain the previous code: the previous code would not truncate non-date parts and hence could store values that were not a multiple of 8640000, which is a violation of the spec.

I think the best thing to do is to put this PR back into draft mode, implement a _full set_ of unit tests for both `Date32` and `Date64` (and their respective builders), and hence ensure we have full clarity on expected behaviour.

Comment (eerhardt) [reply csharp/src/Apache.Arrow/Arrays/Date64Array.cs]: > I disagree with the statement that this is not correct, but I want to understand your viewpoint better...

> From looking at your test case, do you mean to suggest that you believe a conversion to UTC first and then taking the date from the UTC date/time is the most natural way to infer a date from a DateTimeOffset?

What I'm saying is that at this very moment in time, the exact same number of days have passed since the Unix Epoch for me and for you, even though we are in different time zones and our local calendars may be on different days. If there was a program that did:

```C#
var array = new Date32Array.Builder()
                    .Append(DateTimeOffset.Now)
                    .Build();
```

And it ran on both of our machines at the same moment in time, using the proposed changes we could end up with 2 different values in the `array`. That is what is not correct, in my opinion. Since we are taking in a `DateTimeOffset`, we need to respect the `Offset` part of that structure. It is a significant piece of information.

The Unix Epoch represents a moment in time. `DateTimeOffset` represents a moment in time. Taking 2 `DateTimeOffset` instances that are equivalent (for example calling Equals on the above `westernHemisphereTime` and `easternHemisphereTime` returns `true`) should result in the exact same number of days since the Unix Epoch.

Now if a user of this library wants to strip the Offset and set it to `Zero`, then that is up to them. But this library shouldn't be making assumptions that the user doesn't want the Offset taken into account.

Comment (mr-smidge) [reply csharp/src/Apache.Arrow/Arrays/Date64Array.cs]: > ... at this very moment in time, the exact same number of days have passed since the Unix Epoch for me and for you, even though we are in different time zones and our local calendars may be on different days

Ok, I agree with this, because the UNIX Epoch is a point in time rather than a date, and a "day" in this context must therefore be a block of 24 hours rather than a box on a calendar.

Alternatively, if someone said "number of days passed since 1st Jan 1970", then a "day" must be a box on a calendar, because no time element has been mentioned: only dates.  This day count could indeed be different for us in different time zones at the very same moment in time.

I'm happy with the idea that the `DateXX` array builders can accept a `DateTimeOffset` and interpret it as the number of 24-hour blocks since the UNIX Epoch.  I'll ensure this is fully documented, as I don't find this the intuitive interpretation and I'm sure I'm not alone here.

Thank you for clarifying :smile:.

File: csharp/src/Apache.Arrow/Arrays/Date64Array.cs
Comment (mr-smidge) [inline csharp/src/Apache.Arrow/Arrays/Date64Array.cs:70]: An alternative to this code might be:

```c#
// Similar to the code in Date32Array.
return (long)(dateTimeOffset.UtcDateTime.Date - _epochDate).TotalDays * MillisecondsPerDay;
```

Or even:

```c#
// No maths, but requires making another DateTimeOffset only to throw it away again.
return new DateTimeOffset(dateTimeOffset.UtcDateTime.Date, TimeSpan.Zero).ToUnixMilliseconds();
```

@eerhardt , please let me know if you have a preference between any of these.

File: csharp/src/Apache.Arrow/Arrays/Date64Array.cs
Comment (mr-smidge) [reply csharp/src/Apache.Arrow/Arrays/Date64Array.cs]: I've now updated the PR so that treatment of `DateTimeOffset` is as discussed above: it is converted to UTC first and then the days are counted, so that the input is treated as a point in time rather than a calendar date/time.  The docstrings all explain this behaviour (hopefully clearly.

File: csharp/test/Apache.Arrow.Tests/TestDateAndTimeData.cs
Comment (mr-smidge) [inline csharp/test/Apache.Arrow.Tests/TestDateAndTimeData.cs:57]: These properties provide a useful set of test data that acts over the cartesian product of dates/times/kinds/offsets, so that testing all possible combinations is straightforward.

File: csharp/src/Apache.Arrow/Arrays/Date64Array.cs
Comment (eerhardt) [reply csharp/src/Apache.Arrow/Arrays/Date64Array.cs:70]: I'm wondering if we really want this behavior, or if we should just store the `value.ToUnixTimeMilliseconds()` as we were doing before. The code calling this can convert the input to be whole days if it wants.

The reason I think this is because I don't see anything preventing someone from calling `builder.Append(long value)` with a value that isn't a multiple of `86400000`. Looking at Java and C++, I don't see this from being stopped either.

Comment (mr-smidge) [reply csharp/src/Apache.Arrow/Arrays/Date64Array.cs:70]: From what I can tell, the Date64 builder (neither in this PR nor in master) has never had an `Append(long)` method.  One could be added, but that's probably beyond the scope of this PR.  I can see the rationale for not doing a multiple-of-86400000 check on the equivalent method in the C++/Java APIs: the user knows they are building Date64 values, and if they are sophisticated enough to be providing the raw underlying 64-bit type, it seems reasonable to trust them to get it right.

However, if the user is providing a conventional Date-like type, like `DateTime` or `DateTimeOffset`, my view is that the API ought to make it easy to do the right thing and difficult to do the wrong thing:

* I believe we're definitely in agreement on what the Right Thing (TM) is for `DateTime`: the code looks at the year/month/day on it and considers that to be the date.  No ambiguity here.
* We've already found conflicting viewpoints on how to interpret `DateTimeOffset` as a date, based on whether it's treated as an instant in time (take `DateTimeOffset.UtcDateTime.Date`) or a "zoned" date and time (take `DateTimeOffset.Date`).

Regardless of how a date is interpreted, if the Arrow API offers a method accepting `DateTimeOffset` and converting into `Date64`, I feel that it should _never_ violate the spec when doing so:

> Milliseconds (64 bits) indicating UNIX time elapsed since the epoch (no leap seconds), where the values are evenly divisible by 86400000

Keeping the code as `value.ToUnixTimeMilliseconds()` makes it very easy for a user to violate the spec by accident, which is not good.

Having spent a fair bit of time discusing this, I am tempted to _drop_ (or mark obsolete) all builder methods accepting `DateTimeOffset` and only accept `DateTime`, so there is no chance of misinterpretation: it would be up to the caller in possession of a `DateTimeOffset` to call `.Date` or `.UtcDateTime` as per their requirements - similar to something you advocated in an earlier comment in fact!  I'd be keen to hear your thoughts on this...

Comment (eerhardt) [reply csharp/src/Apache.Arrow/Arrays/Date64Array.cs:70]: > From what I can tell, the Date64 builder (neither in this PR nor in master) has never had an Append(long) method.

You're right - sorry I got confused between the different languages' implementations.

I started a conversation on the dev list about this topic, and @wesm had a good idea that we could consider here:

> I think we should validate optionally in ValidateFull in C++. I think to validate unconditionally would be too computationally expensive

https://issues.apache.org/jira/browse/ARROW-9705

I think you've convinced me that the default/right/safe behavior is to ensure the spec is enforced, so ensuring appending a `DateTimeOffset` to a `Date64Array` makes the underlying `long` value divisible by `86400000` is the right thing to do. If necessary, in the future we could add a `long` overload which doesn't do validation in the off chance that someone really did want non-zero time in their `Date64Array`.

Thanks for working through this with me.

**Review by eerhardt — APPROVED** [review]
I'll merge this later today or tomorrow unless I see additional feedback.

Thanks again for your patience with me, @mr-smidge!

**Pull request closed without merging**
